test/functional/ex_cmds/read_spec.lua

```diff
@@ -0,0 +1,93 @@
+local inserted_start = { read_text[1], read_text[2], original_text[1], original_text[2] }
+
+local function test_read(cmd, expected)
+  setline(1, original_text)
+  setcharpos('.', { 0, 0, 0 })
+  execute(cmd)
+  for i, e in ipairs(expected) do
+    eq(e, getline(i))
+  end
+end
+
+local function test_undo(cmd)
+  setline(1, original_text)
+  execute(cmd)
+  execute('undo')
+  eq(original_text, { getline(1), getline(2) })
+end
+
+describe(':read', function()
+  local function cleanup()
+    os.remove(tmp_file)
+  end
+  before_each(function()
+    clear()
+    cleanup()
+    write_file(tmp_file, table.concat(read_text, '\n'), true)
+  end)
+  after_each(cleanup)
+  it('inserts text from file', function()
+    test_read('read ' .. tmp_file, inserted_middle)
+    eq({ 0, 2, 2, 0 }, fn.getpos('.'))
+  end)
+  it('inserts text from shell', function()
+    test_read('read !cat ' .. tmp_file, inserted_middle)
+    eq({ 0, 3, 3, 0 }, fn.getpos('.'))
+  end)
+  it('inserts text from command', function()
+    local make_lines = string.format('let lines="%s"', table.concat(read_text, '\\n'))
+    execute(make_lines)
+    test_read('read :echo lines', inserted_middle)
+  end)
+  it('inserts text from file at specific position', function()
+    test_read('0read ' .. tmp_file, inserted_start)
+    eq({ 0, 1, 2, 0 }, fn.getpos('.'))
+  end)
+  it('inserts text from cmd at specific position', function()
+    test_read('0read !cat ' .. tmp_file, inserted_start)
+    eq({ 0, 2, 3, 0 }, fn.getpos('.'))
+  end)
+  it('inserts text from cmd at specific position', function()
+    local make_lines = string.format('let lines="%s"', table.concat(read_text, '\\n'))
+    execute(make_lines)
+    test_read('0read :echo lines', inserted_start)
+    eq({ 0, 1, 2, 0 }, fn.getpos('.'))
+  end)
+  it('sets fileformat, fileencoding, bomb correctly', function()
+    execute('set fileformat=dos')
+    execute('set fileencoding=latin1')
+    execute('set bomb')
+    execute('read ++edit ' .. tmp_file)
+    eq('fileformat=unix', vim.trim(execute('set fileformat?')))
+    eq('fileencoding=utf-8', vim.trim(execute('set fileencoding?')))
+    eq('nobomb', vim.trim(execute('set bomb?')))
+  end)
+  it('file reads can be undone', function()
+    test_undo('read ' .. tmp_file)
+  end)
+  it('shell reads can be undone', function()
+    test_undo('read !cat ' .. tmp_file)
+  end)
+  it('command reads can be undone', function()
+    local make_lines = string.format('let lines="%s"', table.concat(read_text, '\\n'))
+    execute(make_lines)
+    test_undo('read :echo lines')
+  end)
+  it('errors out correctly when a non-existant file is used', function()
+    eq("Vim(read):E484: Can't open file asdfasdf", pcall_err(execute, ':read asdfasdf'))
+  end)
+  it('errors out correctly when an invalid command is used', function()
+    eq('Vim:E492: Not an editor command: asdfasdf', pcall_err(execute, ':read :asdfasdf'))
```

Comment:
I'm not sure if this error message is the best it could be. Should I have it say `Vim(read):E492: ...`

Alternatively, the error text could be written to the buffer. I don't _think_ it should, but I can see how someone might want that behaviour.

Reply:
> Should I have it say `Vim(read):E492: ...`

if it's not much trouble then I think so.

> Alternatively, the error text could be written to the buffer

`execute()` appears to capture error messages. But for now, let's optimize for implementation simplicity. Can always revisit later if needed.

Reply:
It would actually be marginally simpler to have the error text written to the buffer. That is what would happen if I didn't wrap the vimscript up in a `try` clause. 

When you say we can revisit later on, do you mean before completing the pull request, or some time after this is merged? I imagine this would become a breaking change, so I would want to get it right.

Reply:
I mean after this is merged. That's a good instinct and a good general practice, but sometimes features like this need to see real-world usage before we fully decide the interface. Especially when "the stakes are low", as with this particular behavior decision (error capture vs no error capture).

Errors are not the common case so I think it's better to just optimize for code simplicity here. Thuse removing the try/catch sounds good.

Reply:
Cool. I know it is pretty low stakes, and the best course of action is to see how it goes, but I didn't want to make an assumption.